### ISSUE TITLE
Cluster dies after pods are killed by high memory consumption

### ISSUE THREAD

**Issue by Zelldon** [zeebe-io/zeebe#3181]
**Describe the bug**
In our long running test suite and also in our chaos cluster I saw that one pod was killed by k8s and this cause to break the cluster.

### Long running

 Setup:

 * 8 Partitions
 * 5 Broker
 * Replication factor 3
 * 100 wfi/s
 * 16 worker a 60 jobs

In the long running tests the broker was killed one hour after beginning, due to high memory consumption.

![screen-long-run](https://user-images.githubusercontent.com/2758593/66803208-3f81b300-ef1f-11e9-9f2d-ac8497a8aaa8.png)
![screen-long-run-mem](https://user-images.githubusercontent.com/2758593/66803209-3f81b300-ef1f-11e9-885c-83fb7f5dc193.png)
![screen-long-run-pod](https://user-images.githubusercontent.com/2758593/66803679-a18ee800-ef20-11e9-8eec-01095f9778d2.png)

As we can see the cluster dies for a while and it took around 16 min to come back.
I found the following log in the `zeebe-2` broker.
```
I 2019-10-14T13:09:19.183169569Z 2019-10-14 13:09:19.182 [] [raft-server-system-partition-1] INFO  io.atomix.protocols.raft.impl.RaftContext - RaftServer{system-partition-1} - Transitioning to FOLLOWER
I 2019-10-14T13:23:56.721488421Z 2019-10-14 13:23:56.721 [] [raft-server-system-partition-1] INFO  io.atomix.protocols.raft.impl.RaftContext - RaftServer{system-partition-1} - Found leader 1
I 2019-10-14T13:23:56.882159405Z 2019-10-14 13:23:56.881 [] [raft-server-system-partition-1] INFO  io.atomix.protocols.raft.partition.RaftPartitionGroup - Started
```

The broker 2 was never ready again.

### Chaos cluster

The same I saw on the chaos cluster.

 Setup:

 * 8 Partitions
 * 5 Broker
 * Replication factor 3
 * 300 wfi/s
 * 16 worker a 60 jobs

In the following screenshot we can see one of the pod restarts and that the cluster is not able to process any commands/events.

![screen-chaos-mem](https://user-images.githubusercontent.com/2758593/66804227-347c5200-ef22-11e9-9dc6-284a0b99ec36.png)
![screen-chaos-run](https://user-images.githubusercontent.com/2758593/66804228-347c5200-ef22-11e9-97ca-7cafd05178db.png)

The pods have been restarted quite often. 
![screen-chaos-cluster](https://user-images.githubusercontent.com/2758593/66804226-347c5200-ef22-11e9-9e39-93a8b7ef84ea.png)

At some point the cluster was dead.
![screen-chaos-complete](https://user-images.githubusercontent.com/2758593/66804372-8f15ae00-ef22-11e9-9ae2-bd650fd3cf58.png)

I see no out of memory errors in the log only in kubernetes the memory graph shows me that it reached the limit. In the log I see that at some point it just restarts and prints the starting log (env variables etc.)

**To Reproduce**
I think it is just running the cluster.

**Expected behavior**

 * No restarts :sweat_smile: 
 * No high memory consumption
 * A healthy cluster

**Environment:**
 - OS: k8s
 - Zeebe Version: 0.22.0-SNAPSHOT
 - Configuration: elastic exporter

**Comment by Zelldon** [comment]
In one node of the chaos cluster I see a endless loop of 

```
 
I 2019-10-15T01:22:16.539066152Z 2019-10-15 01:22:16.538 [] [raft-server-raft-atomix-partition-1-state] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService-1 - Restoring local log from position 13559219264728 to 17721039334544
 
I 2019-10-15T01:22:16.539095241Z 2019-10-15 01:22:16.538 [] [raft-server-raft-atomix-partition-6-state] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService-6 - Restoring local log failed at position 12532718651864, retrying.
 
I 2019-10-15T01:22:16.539108614Z java.util.concurrent.CompletionException: io.atomix.cluster.messaging.MessagingException$NoRemoteHandler: No remote message handler registered for this message
	at java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.postComplete(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source) ~[?:?]
	at io.atomix.cluster.messaging.impl.NettyMessagingService.lambda$null$10(NettyMessagingService.java:350) ~[atomix-cluster-3.2.0-alpha6.jar:?]
	at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:30) ~[guava-28.1-jre.jar:?]
	at io.atomix.cluster.messaging.impl.NettyMessagingService.lambda$null$11(NettyMessagingService.java:350) ~[atomix-cluster-3.2.0-alpha6.jar:?]
	at java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.postComplete(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source) ~[?:?]
	at io.atomix.cluster.messaging.impl.AbstractClientConnection$Callback.completeExceptionally(AbstractClientConnection.java:180) ~[atomix-cluster-3.2.0-alpha6.jar:?]
	at io.atomix.cluster.messaging.impl.AbstractClientConnection.dispatch(AbstractClientConnection.java:66) ~[atomix-cluster-3.2.0-alpha6.jar:?]
	at io.atomix.cluster.messaging.impl.AbstractClientConnection.dispatch(AbstractClientConnection.java:39) ~[atomix-cluster-3.2.0-alpha6.jar:?]
	at io.atomix.cluster.messaging.impl.NettyMessagingService$MessageDispatcher.channelRead0(NettyMessagingService.java:827) ~[atomix-cluster-3.2.0-alpha6.jar:?]
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:328) ~[netty-codec-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:302) ~[netty-codec-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1421) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:930) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:794) ~[netty-transport-native-epoll-4.1.39.Final-linux-x86_64.jar:4.1.39.Final]
	at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:424) ~[netty-transport-native-epoll-4.1.39.Final-linux-x86_64.jar:4.1.39.Final]
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:326) ~[netty-transport-native-epoll-4.1.39.Final-linux-x86_64.jar:4.1.39.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:918) ~[netty-common-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[netty-common-4.1.39.Final.jar:4.1.39.Final]
	at java.lang.Thread.run(Unknown Source) [?:?]
Caused by: io.atomix.cluster.messaging.MessagingException$NoRemoteHandler: No remote message handler registered for this message
	... 22 more
 
I 2019-10-15T01:22:16.539386229Z 2019-10-15 01:22:16.538 [] [raft-server-raft-atomix-partition-6-state] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService-6 - Restoring local log from position 12532718651864 to 13460435241168
 
I 2019-10-15T01:22:16.539552900Z 2019-10-15 01:22:16.539 [] [raft-server-raft-atomix-partition-1-state] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService-1 - Restoring local log failed at position 13559219264728, retrying.
 
I 2019-10-15T01:22:16.539574476Z java.util.concurrent.CompletionException: io.atomix.cluster.messaging.MessagingException$NoRemoteHandler: No remote message handler registered for this message
	at java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.postComplete(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source) ~[?:?]
	at io.atomix.cluster.messaging.impl.NettyMessagingService.lambda$null$10(NettyMessagingService.java:350) ~[atomix-cluster-3.2.0-alpha6.jar:?]
	at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:30) ~[guava-28.1-jre.jar:?]
	at io.atomix.cluster.messaging.impl.NettyMessagingService.lambda$null$11(NettyMessagingService.java:350) ~[atomix-cluster-3.2.0-alpha6.jar:?]
	at java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.postComplete(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source) ~[?:?]
	at io.atomix.cluster.messaging.impl.AbstractClientConnection$Callback.completeExceptionally(AbstractClientConnection.java:180) ~[atomix-cluster-3.2.0-alpha6.jar:?]
	at io.atomix.cluster.messaging.impl.AbstractClientConnection.dispatch(AbstractClientConnection.java:66) ~[atomix-cluster-3.2.0-alpha6.jar:?]
	at io.atomix.cluster.messaging.impl.AbstractClientConnection.dispatch(AbstractClientConnection.java:39) ~[atomix-cluster-3.2.0-alpha6.jar:?]
	at io.atomix.cluster.messaging.impl.NettyMessagingService$MessageDispatcher.channelRead0(NettyMessagingService.java:827) ~[atomix-cluster-3.2.0-alpha6.jar:?]
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:328) ~[netty-codec-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:302) ~[netty-codec-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1421) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:930) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:794) ~[netty-transport-native-epoll-4.1.39.Final-linux-x86_64.jar:4.1.39.Final]
	at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:424) ~[netty-transport-native-epoll-4.1.39.Final-linux-x86_64.jar:4.1.39.Final]
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:326) ~[netty-transport-native-epoll-4.1.39.Final-linux-x86_64.jar:4.1.39.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:918) ~[netty-common-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[netty-common-4.1.39.Final.jar:4.1.39.Final]
	at java.lang.Thread.run(Unknown Source) [?:?]
Caused by: io.atomix.cluster.messaging.MessagingException$NoRemoteHandler: No remote message handler registered for this message
	... 22 more
 
I 2019-10-15T01:22:16.539817618Z 2019-10-15 01:22:16.539 [] [raft-server-raft-atomix-partition-1-state] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService-1 - Restoring local log from position 13559219264728 to 17721039334544
 
I 2019-10-15T01:22:16.539837493Z 2019-10-15 01:22:16.539 [] [raft-server-raft-atomix-partition-6-state] DEBUG io.zeebe.distributedlog.impl.DefaultDistributedLogstreamService-6 - Restoring local log failed at position 12532718651864, retrying.
 
I 2019-10-15T01:22:16.539843017Z java.util.concurrent.CompletionException: io.atomix.cluster.messaging.MessagingException$NoRemoteHandler: No remote message handler registered for this message
	at java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.postComplete(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source) ~[?:?]
	at io.atomix.cluster.messaging.impl.NettyMessagingService.lambda$null$10(NettyMessagingService.java:350) ~[atomix-cluster-3.2.0-alpha6.jar:?]
	at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:30) ~[guava-28.1-jre.jar:?]
	at io.atomix.cluster.messaging.impl.NettyMessagingService.lambda$null$11(NettyMessagingService.java:350) ~[atomix-cluster-3.2.0-alpha6.jar:?]
	at java.util.concurrent.CompletableFuture.uniWhenComplete(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.postComplete(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeExceptionally(Unknown Source) ~[?:?]
	at io.atomix.cluster.messaging.impl.AbstractClientConnection$Callback.completeExceptionally(AbstractClientConnection.java:180) ~[atomix-cluster-3.2.0-alpha6.jar:?]
	at io.atomix.cluster.messaging.impl.AbstractClientConnection.dispatch(AbstractClientConnection.java:66) ~[atomix-cluster-3.2.0-alpha6.jar:?]
	at io.atomix.cluster.messaging.impl.AbstractClientConnection.dispatch(AbstractClientConnection.java:39) ~[atomix-cluster-3.2.0-alpha6.jar:?]
	at io.atomix.cluster.messaging.impl.NettyMessagingService$MessageDispatcher.channelRead0(NettyMessagingService.java:827) ~[atomix-cluster-3.2.0-alpha6.jar:?]
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:328) ~[netty-codec-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:302) ~[netty-codec-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1421) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:930) ~[netty-transport-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:794) ~[netty-transport-native-epoll-4.1.39.Final-linux-x86_64.jar:4.1.39.Final]
	at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:424) ~[netty-transport-native-epoll-4.1.39.Final-linux-x86_64.jar:4.1.39.Final]
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:326) ~[netty-transport-native-epoll-4.1.39.Final-linux-x86_64.jar:4.1.39.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:918) ~[netty-common-4.1.39.Final.jar:4.1.39.Final]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[netty-common-4.1.39.Final.jar:4.1.39.Final]
	at java.lang.Thread.run(Unknown Source) [?:?]
Caused by: io.atomix.cluster.messaging.MessagingException$NoRemoteHandler: No remote message handler registered for this message
	... 22 more
 

```

**Comment by Zelldon** [comment]
It is fast reproducible with a load of 300 wfi/s. Did this on the chaos cluster. As you can seen in the screenshot at some point the direct buffer memory spikes to 8 gb and the zeebe 3 node breaks.


![screen-chaos-break](https://user-images.githubusercontent.com/2758593/66818538-c2196b00-ef3d-11e9-8a37-b718753c51fa.png)

At this point in time (11:09) I see again the endless loop I posted above.

**Comment by Zelldon** [comment]
I configured `Xmx and Xms` to be 4 gig and run on the chaos cluster following setup:

 * 1 Partition
 * 3 Brokers
 * 3 replication
 * 3 workers
 * 900 wfi/s

As first all seems good but after some time the heap seem to spike a lot.


![mem-consumption](https://user-images.githubusercontent.com/2758593/66924775-852ba200-f02b-11e9-8aa8-dbe2d17db390.png)


I took a look into the log and it seems that everytime we had a spike elastic exporter was not able to send the bulk.


```
I 2019-10-16T13:05:12.674226162Z 2019-10-16 13:05:12.673 [exporter] [zeebe-2.zeebe.default.svc.cluster.local:26501-zb-fs-workers-5] WARN  io.zeebe.broker.exporter.elasticsearch - Failed to flush at least one bulk request ElasticsearchException[Elasticsearch exception [type=cluster_block_exception, reason=blocked by: [FORBIDDEN/12/index read-only / allow delete (api)];]]
 
I 2019-10-16T13:05:12.674526538Z 2019-10-16 13:05:12.674 [exporter] [zeebe-2.zeebe.default.svc.cluster.local:26501-zb-fs-workers-5] WARN  io.zeebe.broker.exporter.elasticsearch - Failed to flush bulk completely
 
I 2019-10-16T13:05:14.301885369Z 2019-10-16 13:05:14.301 [exporter] [zeebe-2.zeebe.default.svc.cluster.local:26501-zb-fs-workers-7] WARN  io.zeebe.broker.exporter.elasticsearch - Failed to flush at least one bulk request ElasticsearchException[Elasticsearch exception [type=cluster_block_exception, reason=blocked by: [FORBIDDEN/12/index read-only / allow delete (api)];]]
 
I 2019-10-16T13:05:14.301926084Z 2019-10-16 13:05:14.301 [exporter] [zeebe-2.zeebe.default.svc.cluster.local:26501-zb-fs-workers-7] WARN  io.zeebe.broker.exporter.elasticsearch - Failed to flush bulk completely
 
```

**Comment by Zelldon** [comment]
I investigated memory consumption further. The problem why the elastic exporter was not able to send the records was that elastic search was nearly full. If elastic reaches a specific point (Watermark) of disk space it marks indices as read only. That is the reason why we seeing these exceptions above. https://www.elastic.co/guide/en/elasticsearch/reference/current/disk-allocator.html

Currently we catch flush exceptions [here]( 
https://github.com/zeebe-io/zeebe/blob/develop/exporters/elasticsearch-exporter/src/main/java/io/zeebe/exporter/ElasticsearchExporter.java#L93-L101) and logging the cause. This is kind of problematic, since we then just keep going in exporting next records and adding these to the unlimited BulkRequest. This is probably the reason why we seeing these memory spikes where we allocate around 4 gig at once.

We shouldn't catch at this point the exception and let the exporter director do his thing, such that we retry with back off etc. The catch should be in the `flushAndReschedule`, such that we are able to reshedule the next flush.

To get more insights in elastic I added the elasticsearch exporter (prometheus) to our benchmark and to our dashboard the elastic disc usage.

![screen-elastic-usage](https://user-images.githubusercontent.com/2758593/67008569-11040380-f0ea-11e9-8979-386a0fe233f7.png)

**Comment by Zelldon** [comment]
I'm currently a bit unsure with the BulkRequest and the spikes - I mean yes it is not ideal how it currently behaves, but not sure if this is the real issue. I would assume that the GC can't collect the memory when we can't flush the requests. :man_shrugging: **AND** I have spikes also when the indices are writable again. At least until 2 gig.


I simulated this issue with help of Kibana and setting the indices to read only manually.
```
PUT zeebe-record-job/_settings
{
  "index.blocks.read_only_allow_delete": true
}
PUT zeebe-record-workflow-instance/_settings
{
  "index.blocks.read_only_allow_delete": true
}
PUT zeebe-record-deployment/_settings
{
  "index.blocks.read_only_allow_delete": true
}
```

And then we getting these exceptions again, but even if I then disable it the memory usage stays the same (with spikes). I will run the test with more partitions and see how it behaves. Maybe we just need per partition around 2 gigs.

![mem-issue](https://user-images.githubusercontent.com/2758593/67012180-37796d00-f0f1-11e9-99bd-226741d56b86.png)

**Comment by Zelldon** [comment]
It seems to work quite well with the 4 gig max setting. I saw no longer any restarts due to reaching the memory limit nor OOM exceptions. 

![mem-chaos](https://user-images.githubusercontent.com/2758593/67039382-87702800-f121-11e9-8fe7-842fa5017e6d.png)


I also saw during the time the test was running more then once that nodes have been restarted (preemptable nodes) but this seemed not to be an issue for the cluster. In the following image you see some spikes these are the moments where a leader change happend.

At `19:02` we see degrading of performance from 100 wfi/s to 30 wfi/s - exactly the same we seeing in the long-running test suite.

![processing](https://user-images.githubusercontent.com/2758593/67039652-0bc2ab00-f122-11e9-9214-f9a3852f34df.png)


A look in the log also explains that.

```
I 2019-10-17T17:03:34.745531005Z 2019-10-17 17:03:34.745 [] [raft-server-raft-atomix-partition-4] WARN  io.atomix.protocols.raft.roles.LeaderRole - RaftServer{raft-atomix-partition-4}{role=LEADER} - Caught OutOfDiskSpace error! Force compacting logs...
 
I 2019-10-17T17:03:34.745821514Z 2019-10-17 17:03:34.745 [] [raft-server-raft-atomix-partition-4] WARN  io.atomix.protocols.raft.roles.LeaderRole - RaftServer{raft-atomix-partition-4}{role=LEADER} - Caught OutOfDiskSpace error! Force compacting logs...
 ```

Which means for some nodes the disk went full and the leader can't do any progress. One hour later the last nodes are full as well. We also see that there around 10 milion records are not yet exported, which might be the cause, since we then can't delete our log.


![exporting](https://user-images.githubusercontent.com/2758593/67040106-12055700-f123-11e9-9ff3-7ba9d09ceefd.png)

**Comment by Zelldon** [comment]
I think memory wise it looks good now. 

![mem](https://user-images.githubusercontent.com/2758593/67139504-51ca5c80-f251-11e9-88a1-63f531d67a99.png)

I did a test last 24 h on chaos and long running and seems good. Pods on chaos where restarted because of different errors (#3223 )

**Comment by Zelldon** [comment]
Ok after a pod was restarted it went very fast again out of memory. The reason seems to be atomix `FileChannelJournalReader` again and a bit our log replication. 

![usage](https://user-images.githubusercontent.com/2758593/67150871-9ee41800-f2bd-11e9-951f-89257bf2f26c.png)

There a lot of 8 Mb byte arrays which hold by `FileChannelJournalReader`.
Unfortunately I can't upload the hprof file even as zip it is to large.

**Comment by Zelldon** [comment]
There are endless CommandEntries, which cause on restart out of memory.

![usage](https://user-images.githubusercontent.com/2758593/67375240-a12ac880-f582-11e9-81b1-a62fb1257d5f.png)

**Comment by npepinpe** [comment]
Using the mapped storage instead of disk would mitigate this issue, I think, as the `MappedJournalSegmentWriter` does not need to allocate/keep a buffer of two entries.

**Comment by Zelldon** [comment]
Would be related to that https://github.com/zeebe-io/zeebe/issues/3094

**Comment by deepthidevaki** [comment]
Observed similar behavior today. Pods just died at the peak memory usage. 
![image](https://user-images.githubusercontent.com/1997478/68685121-e894eb00-0569-11ea-8dcf-9e811d593ec7.png)

**Comment by Zelldon** [comment]
After the latest fixes the processing looks really good but still we get some OOM Errors. Fortunately the cluster heals after the restart of the node.

![oom](https://user-images.githubusercontent.com/2758593/69433355-c22a3900-0d3b-11ea-950d-5f4ed4c69b22.png)

**Comment by Zelldon** [comment]
The latest OOM's were caused because we introduced the back pressure in the log appender, where we reduced the load in atomix. This means atomix is now able to do more often snapshots (every 15 s). After doing a snapshot the snapshot is directly installed and the primitive recreated. This caused several issues because listeners and sessions were not correctly cleaned up. With the latest fixes (e.g. https://github.com/zeebe-io/atomix/pull/87) this should be no longer an issue.
